controllers/idle_handlers.go

```diff
@@ -110,11 +83,132 @@ func (r *ImageBasedUpgradeReconciler) handleFinalizeFailure(ctx context.Context,
 	return doNotRequeue(), nil
 }
 
+// cleanup cleans stateroots, precache, backup, ibu files and OADP operator
+// returns true if all cleanup tasks were successful
+func (r *ImageBasedUpgradeReconciler) cleanup(
+	ctx context.Context, allUnbootedStateroots bool,
+	ibu *lcav1alpha1.ImageBasedUpgrade) (bool, string) {
+	// try to clean up as much as possible and avoid returning when one of the cleanup tasks fails
+	// successful means that all the cleanup tasks completed without any error
+	successful := true
+	errorMessage := ""
+
+	var handleError = func(err error, msg string) {
+		successful = false
+		r.Log.Error(err, msg)
+		errorMessage += msg + " "
+	}
+
+	if err := r.cleanupStateroots(allUnbootedStateroots, ibu); err != nil {
+		handleError(err, "failed to cleanup stateroots.")
+	}
+	if err := r.Precache.Cleanup(ctx); err != nil {
+		handleError(err, "failed to cleanup precaching resources.")
+	}
+	if allRemoved, err := r.BackupRestore.CleanupBackups(ctx); err != nil {
+		handleError(err, "failed to cleanup backups.")
+	} else if !allRemoved {
+		err := errors.New("failed to delete all the backup CRs.")
+		handleError(err, err.Error())
+	}
+	if err := r.BackupRestore.DeleteOadpOperator(ctx, backuprestore.OadpNs); err != nil {
+		handleError(err, "failed to delete OADP operator.")
+	}
+	if err := cleanupIBUFiles(); err != nil {
+		handleError(err, "failed to cleanup ibu files.")
+	}
+
+	return successful, errorMessage
+}
+
+// cleanupStateroot cleans all unbooted stateroots or desired stateroot
+// depending on allUnbootedStateroots argument
+func (r *ImageBasedUpgradeReconciler) cleanupStateroots(
+	allUnbootedStateroots bool, ibu *lcav1alpha1.ImageBasedUpgrade) error {
+	if allUnbootedStateroots {
+		return r.cleanupUnbootedStateroots()
+	}
+	return r.cleanupUnbootedStateroot(getDesiredStaterootName(ibu))
+}
+
 func cleanupIBUFiles() error {
-	if _, err := os.Stat(common.PathOutsideChroot(utils.IBUWorkspacePath)); err == nil {
-		if err := os.RemoveAll(filepath.Join(utils.Host, utils.IBUWorkspacePath)); err != nil {
-			return fmt.Errorf("removing %s failed: %w", utils.IBUWorkspacePath, err)
+	if _, err := os.Stat(common.PathOutsideChroot(utils.IBUWorkspacePath)); err != nil {
+		return nil
+	}
+	if err := os.RemoveAll(common.PathOutsideChroot(utils.IBUWorkspacePath)); err != nil {
+		return fmt.Errorf("removing %s failed: %w", utils.IBUWorkspacePath, err)
+	}
+	return nil
+}
+
+func (r *ImageBasedUpgradeReconciler) cleanupUnbootedStateroots() error {
+	status, err := r.RPMOstreeClient.QueryStatus()
+	if err != nil {
+		return err
+	}
+
+	bootedStateroot := ""
+	staterootsToRemove := make([]string, 0)
+	// since undeploy shifts the order, undeploy in the reverse order
+	for i := len(status.Deployments) - 1; i >= 0; i-- {
+		deployment := &status.Deployments[i]
+		if deployment.Booted {
+			bootedStateroot = deployment.OSName
+			continue
+		}
+		if err := r.OstreeClient.Undeploy(i); err != nil {
```

Comment:
This is actually undeploying all unbooted deployments, not just those in unbooted stateroot(s). You need to create the list of stateroots to be deleted first (or determine the booted stateroot). Then you can check that deployment to see if it's in the booted stateroot, as is done in the cleanupUnbootedStateroot function below.

Why not have this function just collect the list of unbooted stateroots, then call cleanupUnbootedStateroot for each?

Reply:
> Why not have this function just collect the list of unbooted stateroots, then call cleanupUnbootedStateroot for each?

why though? that function query and iterates the deployments again. not efficient?

What is wrong with undeploying an unbooted deployment? (apart from name of the function)

Reply:
this is called only in finalize. how could another deployment be in the stateroot we just created?

Reply:
Calling the other function removes code duplication. The difference between the two is one deletes all unbooted stateroots, the other just a specified stateroot. I could ask why not just have the one function that deletes all unbooted stateroots. When would you need to differentiate deleting one vs multiple? Why have two separate functions?

As for the behaviour here, I assume deleting secondary deployments from the booted stateroot was not the intent, and is therefore a bug. Being ok with the bug's behaviour doesn't mean it shouldn't be addressed.